Bug 1896446: fix configuring git when global proxy exists

gabemontero opened the pull request:
/assign @adambkaplan 

@otaviof @coreydaley while we were discussing this in scrum another customer case with build logs including trace came in ... so of course we learned the problem was none of the potential issues we discussed during scrum

adambkaplan commented on pkg/build/builder/cmd/scmauth/util.go:
```diff
@@ -59,3 +60,41 @@ func ensureGitConfigIncludes(path string, context SCMAuthContext) error {
 	return ioutil.WriteFile(gitconfig, content, 0600)
 }
+
+func EnsureGitConfigIncludesProxyConfig(context SCMAuthContext, gitSource *buildv1.GitBuildSource) error {
```
Need a unit test verifying the behavior of this function.

gabemontero replied:
unit test pushed @adambkaplan

adambkaplan reviewed (changes requested):
/hold

We need to ensure that we don't regress on this feature - in particular:

1. E2e tests that run git clone using source secrets. I've started discussions with @stevekuznetsov and @bparees on this.
2. We need to run our existing build suite behind a proxy on relevant PRs. I believe there is a conformance suite that runs behind proxy today, but the full build suite doesn't.

@gabemontero - does @smarterclayton's efforts to run CI as disconnected/mirrored clusters also include proxy configurations?

adambkaplan commented on pkg/build/builder/cmd/scmauth/util_test.go:
```diff
@@ -0,0 +1,38 @@
+	"testing"
+)
+
+func TestEnsureGitConfigIncludesProxyConfig(t *testing.T) {
```
I think the following scenarios need to be tested:

1. Proxy variables are not defined
2. `GIT_CONFIG` is not set (verify error returned)
3. Permutations of proxy variables:
   1. HTTPS_PROXY only set
   2. HTTP_PROXY only set
   3. Both set

gabemontero replied:
additional test case scenarios pushed @adambkaplan

adambkaplan reviewed (changes requested):
Being a stickler on the unit test - want to ensure our reasonable code branches are covered.

gabemontero commented:
> /hold
> 
> We need to ensure that we don't regress on this feature - in particular:
> 
>     1. E2e tests that run git clone using source secrets. I've started discussions with @stevekuznetsov and @bparees on this.
> 
>     2. We need to run our existing build suite behind a proxy on relevant PRs. I believe there is a conformance suite that runs behind proxy today, but the full build suite doesn't.
> 

I've issued a `test e2e-builds openshift/builder#187 aws,proxy` to cluster-bot and it says it is off and doing it ... we'll see how it goes

> 
> @gabemontero - does @smarterclayton's efforts to run CI as disconnected/mirrored clusters also include proxy configurations?

I've seen nothing to indicate that.  Don't know enough about the disconnected env, but it seems like a bit of a stretch that it can tolerate inserting a proxy between say the cluster and the mirror or the cluster and SCMs also hosted within on its inititial foray.  More of a future item.  But of course reach out to Clayton and discuss.

gabemontero commented:
> > /hold
> > We need to ensure that we don't regress on this feature - in particular:
> > ```
> > 1. E2e tests that run git clone using source secrets. I've started discussions with @stevekuznetsov and @bparees on this.
> > 
> > 2. We need to run our existing build suite behind a proxy on relevant PRs. I believe there is a conformance suite that runs behind proxy today, but the full build suite doesn't.
> > ```
> 
> I've issued a `test e2e-builds openshift/builder#187 aws,proxy` to cluster-bot and it says it is off and doing it ... we'll see how it goes
> 

already found a problem while re-examining my changes  ... will push new commit ... let's see if the above test invocation catches the problem I just found :-)

gabemontero commented:
ok update pushed ... it had to do with a nil gitClient if there was no git source ref ... the gitClient looked like it was still used down the line in that case 

I'm curious if that permutation is caught in regular tests or proxy tests ... we'll see

gabemontero commented:
So @adambkaplan  my `test e2e-builds openshift/builder#187 aws,proxy`

had a bunch of failed tests with this in the build log:

```
I1111 20:47:54.261595       1 repository.go:434] Executing git config --global --add http.proxy http://redacted@35.196.128.173:3128/
I1111 20:47:54.269900       1 repository.go:517] Error executing command: exit status 128
F1111 20:47:54.306544       1 helpers.go:115] error: cannot setup global proxy: fatal: $HOME not set
```

full run is at https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1326613430488010752

there's a *chance* that since this run ran before my git client init fix that I noted https://github.com/openshift/builder/pull/187#issuecomment-725644457 has some bearing so I'm retrying

but if we see the same thing again, I think that could mean moving off of the global git config route even in the non-source secret case .... though I'm wondering why this was not seen in the 2 customer logs and the testing you did last go around in 4.6 ... maybe the cluster-bot proxy setup has some differences ???

Will report back the next run and will go from there.

gabemontero commented:
seeing it in the e2e direct from this PR as well:

```
fail [github.com/openshift/origin/test/extended/builds/cluster_config.go:330]: Expected
    <string>: error: cannot setup global proxy: fatal: $HOME not set
to contain substring
    <string>: Could not resolve proxy: invalid.proxy.redhat.com
```

I think I can borrow from the unit tests and set it up locally all the time, even when there are no source secrets

will start on that Thursday

gabemontero commented:
getting a little closer with the latest commit that moves off of global config, but still not quite there
- I was failing to initialize the SCMAuthContext in the non source secret case
- that said, I've now realized that the other SCMAuth's do not necessarily update `.gitconfig`, in fact only 1 of the 4 do, so I need to rework things so my new ensure proxy method handles the setup / reuse of `.gitconfig` as needed

gabemontero commented:
> getting a little closer with the latest commit that moves off of global config, but still not quite there
> 
>     * I was failing to initialize the SCMAuthContext in the non source secret case
> 
>     * that said, I've now realized that the other SCMAuth's do not necessarily update `.gitconfig`, in fact only 1 of the 4 do, so I need to rework things so my new ensure proxy method handles the setup / reuse of `.gitconfig` as needed

OK pushed / squashed these change ... will see how the e2e's here react, and if all good, will kick off the cluster-bot e2e-builds with proxy run

gabemontero commented:
various compromised env looking flakes in image-eco https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws-image-ecosystem/1326914847996121088

will retest later and see if any consistency arises

gabemontero commented:
cluster install could not get off the ground https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws/1326914847845126144

gabemontero commented:
more unhealthy env in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws-builds/1326914847929012224

littered with cannot create kube client for test and default SA not ready

beginning to see some issues in other PRs as well

gabemontero commented:
`   * could not run steps: step e2e-aws failed: failed to acquire lease: resources not found `

gabemontero commented:
`   * could not run steps: step e2e-aws-image-ecosystem failed: failed to acquire lease: resources not found `

gabemontero commented:
the e2e-aws-builds run 
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws-builds/1326953179845234688 
got off the ground 

it has some failures, but none of them are related to the git phase of the build ... in fact in some of the failures, the dump of the build shows the git clone is good

also, none of the global `git config` failed because of no HOME like before, so we're past that

the failures seem to be lower level in the stack or env issues ... a fair amount of endpoints associated with deployments (where came up successfully, and which were triggered as a result of successful builds) not available in time

I've launched via cluster-bot another e2e-builds suite against this PR using a proxy enabled cluster

I'll post the link / analysis when it completes

gabemontero commented:
the proxy cluster test https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1326983572099174400 

had a failure with the test [sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via cluster configuration build config no ocm rollout Apply git proxy configuration to build pod 

a build that was suppose to fail did not

will need to debug that

gabemontero commented:
I pushed a change to fix the format of proxies with .gitconfig to mirror what `git config --global ... ` does (turns out that is the one that updates the .gitconfig in the home dir ... at least I learned something)

`sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via cluster configuration build config no ocm rollout Apply git proxy configuration to build pod`

is still failing ... both here in this PR and the in the cluster-bot e2e-builds runs in a proxy env

time to add some debug

gabemontero commented:
ok in adding debug I discovered I was not fully propagating the HOME and GIT_CONFIG env vars in the global proxy / no private repo case

debug plus that fix pushed ... let's see how `[sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via cluster configuration build config no ocm rollout Apply git proxy configuration to build pod` does

gabemontero commented:
cluster-bot e2e-builds with proxy came back ... close, but no cigar:

```
    GGM create git config content [http]
    proxy = http://invalid.proxy.redhat.com:3288
    [https]
    .proxy = https://invalid.proxy.redhat.com:3288
    
    GGM setup git env no source secret envs after .gitconfig create are GIT_ASKPASS=true,HOME=/tmp/git935033504,GIT_CONFIG=/tmp/git935033504/.gitconfig
    Cloning "https://github.com/sclorg/nodejs-ex" ...
    error: fatal: bad config line 4 in file /tmp/git935033504/.gitconfig
to contain substring
    <string>: Could not resolve proxy: invalid.proxy.redhat.com
```

*ONE PERIOD OFF*  with `.proxy = https://invalid.proxy.redhat.com:3288`:-)

will have new commit up in a minute

gabemontero commented:
/retest

gabemontero commented:
OK we got a passing run of the proxy test in question:

`[sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via  cluster configuration  build config no ocm rollout Apply git proxy  configuration to build pod | 1m58s`

so progress 

a bunch of other tests failed cause the cluster got fubared and new projects did not get their default SA's in time

gabemontero commented:
the 

`:[sig-builds][Feature:Builds][Slow] builds should support proxies start build with broken proxy and a no_proxy override should start a docker build and wait for the build to succeed`

and 

`[sig-builds][Feature:Builds][Slow] builds should support proxies start build with broken proxy and a no_proxy override should start an s2i build and wait for the build to succeed expand_more`

failed in the latest e2e-builds with proxy via cluster-bot run .... looking at those next

gabemontero commented:
with the last e2e-aws-builds run only 2 of the proxy related e2e's failed and that was because of debug I had added ... it *appears* the approaches I found on stackoverflow etc. for representing no_proxy in  .gitconfig files worked

removed the debug ... let's see if anything past that point in those 2 remaining proxy related tests crop up

gabemontero commented:
ok all the proxy tests passed in e2e-aws-builds ... the failures were whatever current regression in 4.7 is hitting things beneath use ... search.ci shows the same errors in other PRs

gonna start a cluster-bot e2e-aws-builds run now

then I'm gonna a push a change which removes setting NO_PROXY on the git command's envs and confirm that has no bearing

gabemontero commented:
Only only 1 of test case failures on the cluster-bot e2e-builds on a aws proxy cluster run at https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1327820360623591424

was proxy related:  ` [sig-builds][Feature:Builds][Slow] builds should support proxies start build with broken proxy and a no_proxy override should start a docker build and wait for the build to succeed expand_more`

However, that appears to be a false negative, with some *very* odd timestamps in the build log.  And perhaps overall lack of health in this cluster.

So the abort on waiting for the build is noted here:

```
Nov 15 04:31:12.713: INFO: Waiting for sample-docker-build-noproxy-1 to complete

Nov 15 04:41:13.955: INFO: WaitForABuild returning with error: timed out waiting for the condition
Nov 15 04:41:13.955: INFO: Done waiting for sample-docker-build-noproxy-1: util.BuildResult{BuildPath:"build.build.openshift.io/sample-docker-build-noproxy-1", BuildName:"sample-docker-build-noproxy-1", StartBuildStdErr:"", StartBuildStdOut:"build.build.openshift.io/sample-docker-build-noproxy-1", StartBuildErr:error(nil), BuildConfigName:"", Build:(*v1.Build)(0xc001691180), BuildAttempt:true, BuildSuccess:false, BuildFailure:false, BuildCancelled:false, BuildTimeout:true, LogDumper:(util.LogDumperFunc)(nil), Oc:(*util.CLI)(0xc0015ea480)}
 with error: Timed out waiting for build "sample-docker-build-noproxy-1" to complete
```

The 10 minute difference of 04:31 to 04:41 is the expected wait duration.

However, when the test's diagnostic kicks in and then dump the build's log, it completed successfully, with the git clone of the ruby repo of course working.

But check this oddity in the build log when trying to get the base image:

```
2020-11-15T04:31:42.936208180Z Getting image source signatures
2020-11-15T04:31:43.804646613Z Copying blob sha256:d518f7038c56d881774a7a29c5eb0c5dab9a9c5223a989b818ec4f2ca864867c
2020-11-15T04:31:43.866126803Z Copying blob sha256:3811c17826fed41d93a6e6bcfb6a048828f5f72e3032bb61e8f7cd0785995bfc
2020-11-15T04:31:43.870853258Z Copying blob sha256:58ef20b02aca8ea0795927176625aef3850209d126949927908234c97bc3c2c4
2020-11-15T04:31:43.898022572Z Copying blob sha256:2b63304f1869b42027cc113af46692a963e05ef7867c88d36ef9c75ffc2fdfc7
2020-11-15T04:31:46.263860183Z Copying blob sha256:84c2003778cc8a5ee85c049e93bb921cb5cb9868b9ed2b0a6652504730d063fe
2020-11-15T05:23:47.728988336Z Warning: Pull failed, retrying in 5s ...
2020-11-15T05:23:52.729157076Z Asked to pull fresh copy of "registry.redhat.io/rhscl/ruby-25-rhel7".
2020-11-15T05:23:52.729229681Z I1115 05:23:52.729172       1 daemonless.go:165] looking for config.json at /var/run/secrets/openshift.io/pull/config.json
2020-11-15T05:23:52.729247120Z I1115 05:23:52.729229       1 cfg.go:151] error reading file: open /var/run/secrets/openshift.io/pull/config.json: no such file or directory
2020-11-15T05:23:52.729345395Z I1115 05:23:52.729268       1 daemonless.go:165] looking for .dockerconfigjson at /var/run/secrets/openshift.io/pull/.dockerconfigjson
2020-11-15T05:23:52.729381070Z I1115 05:23:52.729351       1 cfg.go:151] error reading file: open /var/run/secrets/openshift.io/pull/.dockerconfigjson: no such file or directory
2020-11-15T05:23:52.729381070Z I1115 05:23:52.729364       1 daemonless.go:165] looking for .dockercfg at /var/run/secrets/openshift.io/pull/.dockercfg
2020-11-15T05:23:52.729594182Z I1115 05:23:52.729550       1 daemonless.go:165] found valid .dockercfg at /var/run/secrets/openshift.io/pull/.dockercfg
2020-11-15T05:23:52.729970187Z I1115 05:23:52.729821       1 config.go:110] looking for .dockercfg at /var/run/secrets/openshift.io/pull/.dockercfg
2020-11-15T05:23:52.730190241Z I1115 05:23:52.730120       1 config.go:125] found .dockercfg at /var/run/secrets/openshift.io/pull/.dockercfg
2020-11-15T05:23:56.556696665Z Getting image source signatures
2020-11-15T05:23:56.562065421Z Copying blob sha256:d518f7038c56d881774a7a29c5eb0c5dab9a9c5223a989b818ec4f2ca864867c
2020-11-15T05:23:56.961415569Z Copying blob sha256:3811c17826fed41d93a6e6bcfb6a048828f5f72e3032bb61e8f7cd0785995bfc
2020-11-15T05:23:57.536660031Z Copying blob sha256:84c2003778cc8a5ee85c049e93bb921cb5cb9868b9ed2b0a6652504730d063fe
2020-11-15T05:24:00.114475015Z Copying blob sha256:2b63304f1869b42027cc113af46692a963e05ef7867c88d36ef9c75ffc2fdfc7
2020-11-15T05:24:00.471557672Z Copying blob sha256:58ef20b02aca8ea0795927176625aef3850209d126949927908234c97bc3c2c4
2020-11-15T05:24:00.471855569Z Copying config sha256:2b3662a998cc53ea6cee3a6067ad3fad9873eef81097466c2a579ee08f7be9c7
2020-11-15T05:24:00.473983264Z Writing manifest to image destination
2020-11-15T05:24:00.475561896Z Storing signatures

```

There is a hiccup trying to get it sorted, a 50 minute or so gap, and then the `2020-11-15T05:23:52.729157076Z Asked to pull fresh copy of "registry.redhat.io/rhscl/ruby-25-rhel7".`

In any event, my opinion at least @adambkaplan , none of that weirdness is releated to this change, and I *think* we now have validations of this change wrt the current set of proxy related tests in the suite both with and without proxy configured.

I'll be issuing more test runs today to see if I can get any repeat validations, but I think when we come in on Monday, the next round of code review is warranted, as well as seeing about how to engage with the rest of the org to sort out 4.7 health (looking at this stuff this weekend, I'm not really sure yet where to direct any bugzillas around the 4.7 instability).

/retest

gabemontero commented:
son of a gun ... we got a **PASSING** e2e-aws-builds :-) at https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws-builds/1327985351083429888

there you go re: unrelated irregularity speculation in https://github.com/openshift/builder/pull/187#issuecomment-727580444

gabemontero commented:
/retest

adambkaplan commented:
/retest

adambkaplan commented on pkg/build/builder/cmd/scmauth/util.go:
```diff
@@ -7,19 +7,43 @@ import (
+		if gitSource.HTTPSProxy != nil && len(*gitSource.HTTPSProxy) > 0 {
+			content = content + fmt.Sprintf("[https]\nproxy = %s\n", *gitSource.HTTPSProxy)
+		}
+		if gitSource.NoProxy != nil && len(*gitSource.NoProxy) > 0 {
+			noProxies := strings.Split(*gitSource.NoProxy, ",")
+			for _, noProxy := range noProxies {
+				if strings.Contains(gitSource.URI, noProxy) {
+					content = content + fmt.Sprintf("[http \"%s\"]\nproxy =\n", gitSource.URI)
+					content = content + fmt.Sprintf("[https \"%s\"]\nproxy =\n", gitSource.URI)
+				}
+			}
+		}
```
We could use a comment to explain what we're doing here.

Ex: "To set up NO_PROXY in a .gitconfig file, add `http` and `https` records for the specific URI that unset `proxy`"

gabemontero replied:
part of next push

adambkaplan commented on pkg/build/builder/cmd/scmauth/util.go:
```diff
@@ -7,19 +7,43 @@ import (
+			noProxies := strings.Split(*gitSource.NoProxy, ",")
+			for _, noProxy := range noProxies {
+				if strings.Contains(gitSource.URI, noProxy) {
+					content = content + fmt.Sprintf("[http \"%s\"]\nproxy =\n", gitSource.URI)
```
Do we add the full URI here, or just the hostname?

gabemontero replied:
url in the sense that you include the http:// or https:// .... from https://superuser.com/questions/719807/how-can-i-exclude-hostnames-for-using-the-default-proxy-in-git

```
[http]
    #sslVerify = false
    proxy = http://proxy.corporate.net:8080
[http "http://ourGitRepository.corporate.net/"]
    #sslVerify = false
    proxy = 
```

note, we have specific no-proxy tests in our e2e's, where we put an invalid proxy address, but the no-proxy setting still allows the git clone to work

adambkaplan commented on pkg/build/builder/cmd/scmauth/util.go:
```diff
@@ -7,19 +7,43 @@ import (
+		if gitSource.NoProxy != nil && len(*gitSource.NoProxy) > 0 {
+			noProxies := strings.Split(*gitSource.NoProxy, ",")
+			for _, noProxy := range noProxies {
+				if strings.Contains(gitSource.URI, noProxy) {
```
Can we convert the URI to a net/url and test just the hostname portion? https://golang.org/pkg/net/url/#URL

gabemontero replied:
I'll give it a try in a separate commit and assuming it works, we can iterate over it

gabemontero replied:
ok separate commit pushed ... will report back after I've analyzed the pr's e2e's and cluster-bot with proxy e2e

adambkaplan commented on pkg/build/builder/cmd/scmauth/util.go:
```diff
@@ -3,23 +3,53 @@ package scmauth
+			content = content + fmt.Sprintf("[https]\nproxy = %s\n", *gitSource.HTTPSProxy)
+		}
+		if gitSource.NoProxy != nil && len(*gitSource.NoProxy) > 0 {
+			gitURL, err := url.Parse(gitSource.URI)
+			if err != nil {
+				return err
+			}
+			noProxies := strings.Split(*gitSource.NoProxy, ",")
+			for _, noProxy := range noProxies {
+				if strings.HasSuffix(gitURL.Host, noProxy) {
```
I may have led you astray here. Will this error out for an SCP-style URI? ex: `git@github.com/...`

gabemontero replied:
Yeah it does not error out, but it does not do what we want.  The host field is blank.  From goland playground:

```
package main

import (
	"fmt"
	"log"
	"net/url"
)

func main() {
	u, err := url.Parse("git@github.com/openshift/origin")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("string")
	fmt.Println(u.String())
	fmt.Println("just host")
	fmt.Println(u.Host)
}

```
string
git@github.com/openshift/origin
just host


```

```

gabemontero replied:
per conversation in slack ... I'll push up a commit soon that checks the scheme to see if http or https so as to bypass git@ or git:  (which both are ssh / key related).

going to wait until the current e2e-*-builds runs are complete first though ... just curious

gabemontero replied:
ok all the proxy related tests in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/187/pull-ci-openshift-builder-master-e2e-aws-builds/1328407467607986176 passed 

that said, I'll push up the new form in a moment @adambkaplan

gabemontero commented:
ok @adambkaplan checkout https://github.com/openshift/builder/commit/cfb0631be673687e5c2f20725fa5727c65a0396e when you get cycles next

gabemontero commented:
hmmm ...` [sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via cluster configuration build config no ocm rollout Apply git proxy configuration to build pod` failed again with this new URL form with the latest e2e-aws-builds after several passing runs @adambkaplan 

I'll investigate more on Tuesday .... then see if tweaking this new commit some more is needed or dropping it altogether is needed

gabemontero commented:
well, so far, while `[sig-builds][Feature:Builds][Serial][Slow][Disruptive] alter builds via cluster configuration build config no ocm rollout Apply git proxy configuration to build pod` is now failing consistently in CI with the new commit based on @adambkaplan 's suggestion, my debug to date in CI does not explain why, and when I run the scenario locally with my changes, it fails as expected, and the debug I added shows the new setup looks good.

I suppose reverting the commit is still an option, but this varying behavior now concerns me.

Gonna add one more round of additional debug ... based on what is shows, go from there.

gabemontero commented:
gotta theory / potential fix with some more debug to validate

adambkaplan commented:
/close

Going with my approach in #189 instead
